File: leetcode/jumpGameIII.js
// Leetcode 1306: Jump Game III
/**
 * @param {number[]} arr
 * @param {number} start
 * @return {boolean}
 */
var canReach = function(arr, idx) {
  
  let set = new Set();
  
  function reach(arr, idx) {
    if (idx >= 0 && idx <= arr.length - 1) {
      if (arr[idx] === 0) {
          return true;
      } 
    } else {
      return false;
    }
    
    if (set.has(idx)) {
      return false;
    } else {
      set.add(idx);
    }

    return reach(arr, idx + arr[idx]) || reach(arr, idx - arr[idx]);
  }
  return reach(arr, idx);
};


let arr = [4,2,3,0,3,1,2];
console.log(canReach(arr, 0));

File: leetcode/jump_game_iii.py
# Leetcode 1306. Jump Game III
from typing import List
class Solution:
  def canJump(self, nums: List[int], start: int) -> int:
    n = len(nums)
    visited = set()

    def backtrack(i):
      if i < 0 or i >= n:
        return False
      if nums[i] in visited:
        return False
      if nums[i] == 0:
        return True
      
      visited.add(i)
      return backtrack(i + nums[i]) or backtrack(i - nums[i])
  
    return backtrack(start)

if __name__ == "__main__":
  sol = Solution()
  arr = [3,0,2,1,2]
  print(sol.canJump(arr, 5))